enhancements/network/ovn-k-localnet-api.md

```diff
@@ -0,0 +1,507 @@
+        "physicalNetworkName: "tenantblue",
+        "subnets": "192.168.0.0/16,2001:dbb::/64",
+        "allowPersistentIPs": true,
+        "excludeSubents: "10.100.50.0/24",
+        "mtu": 9000,
+        "vlanID": 200
+    }'
+```
+
+#### Go types
+The CUDN `spec.network.topology` field should be extended to accept `Localnet` string.
+And the `spec.network` struct `NetworkSpec` should be to have additional field for the localnet topology  configuration:
+```go
+const NetworkTopologyLocalnet NetworkTopology = "Localnet"
+
+...
+
+// NetworkSpec defines the desired state of UserDefinedNetworkSpec.
+// +union
+type NetworkSpec struct {
+    // Topology describes network configuration.
+    //
+    // Allowed values are "Layer3", "Layer2" and "Localnet".
+    // Layer3 topology creates a layer 2 segment per node, each with a different subnet. Layer 3 routing is used to interconnect node subnets.
+    // Layer2 topology creates one logical switch shared by all nodes.
+    // Localnet topology attach to the overlay local network. Enables egress to the provider's physical network. 
+    //
+    // +kubebuilder:validation:Required
+    // +required
+    // +unionDiscriminator
+    Topology NetworkTopology `json:"topology"`
+    
+    ...
+    
+    // Localnet is the Localnet topology configuration.
+    // +optional
+    Localnet *LocalnetConfig `json:"localnet,omitempty"`
+}
+```
+
+The CUDN spec should have additional validation rule for `spec.network.topology` field:
+```go
+// ClusterUserDefinedNetworkSpec defines the desired state of ClusterUserDefinedNetwork.
+type ClusterUserDefinedNetworkSpec struct {
+    ...
+    // +required
+    Network NetworkSpec `json:"network"`
+}
+```
+
+As of today the CUDN has validation to ensure spec.network is immutable:
+```go
+// +kubebuilder:validation:XValidation:rule="self == oldSelf", message="Network spec is immutable"
+```
+This validation should be changed to allow mutation localnet topology configurations, at least MTU and VLAN.
+
+Localnet topology configuration type 
+```go
+type LocalnetConfig struct {
+    // Role describes the network role in the pod.
+    //
+    // Allowed value is "Secondary".
+    // Secondary network is only assigned to pods that use `k8s.v1.cni.cncf.io/networks` annotation to select given network.
+    //
+    // +kubebuilder:validation:Required
+    // +required
+    Role NetworkRole `json:"role"`
+    
+    // PhysicalNetworkName the OVS bridge-mapping's network-name configured in the nodes.
+    //
+    // +kubebuilder:validation:Required
+    // +required
+    PhysicalNetworkName string `json:"name"`
+    
+    // Subnets are used for the pod network across the cluster.
+    // Dual-stack clusters may set 2 subnets (one for each IP family), otherwise only 1 subnet is allowed.
+    //
+    // The format should match standard CIDR notation (for example, "10.128.0.0/16").
+    // This field must be omitted if `ipam.mode` is `Disabled`.
```

Comment:
How do you plan to enforce this?

Reply:
I would recommend including any CEL validations that you intend to have for this included in some way for reviewers of the document. Whether that is explicitly in this proposed API or by placing a link to similar existing CEL validations you intend to put in place when creating the actual PR to the openshift/api repository.

Reply:
I added the additional validations to the doc, please see https://github.com/openshift/enhancements/pull/1750/files#diff-65ff4f11b53425ec17e6fbd228843c58470c309f5031c38c345010775bb378d9R423-R425